### PR TITLE
Added delete_by_property method

### LINKED ISSUE #42565: Add delete_by_property method in weaviate hook
### Description

Currently `delete_collections` is implemented in weaviate hook which deletes all or specific collections if collection_names are provided.
Delete collections by_property method would allow users to delete by providing the property names and deletion of objects that match a set of criteria.
This feature was added in Weaviate v1.21.

example:
```
collection_names = ["collection_a", "collection_b", "collection_c"]
by_property = ["question", "answer", "category"]
weaviate_hook.delete_collections(collection_names=collection_names, by_property=by_property)
```

### Use case/motivation

By adding the delete_by_property method, users can delete collections based on various filtering criteria, making data management more targeted and efficient.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by harjeevanmaan**
- Delete collections by specifying one or more properties.
- Use optional filters (`contains_any` or `contains_all`) to  allow deletion of objects that match a set of criteria.
-  Fixes #42565

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 1]
```diff
@@ -190,7 +190,12 @@ def get_collection(self, name: str) -> Collection:
         return client.collections.get(name)
 
     def delete_collections(
-        self, collection_names: list[str] | str, if_error: str = "stop"
+        self,
+        collection_names: list[str] | str,
+        by_property: list[str] | str | None = None,
+        contains_any: list[Any] | str | None = None,
+        contains_all: list[Any] | str | None = None,
+        if_error: str = "stop",
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py]:
```suggestion
        if_error: str = "stop",
        *,
        by_property: list[str] | str | None = None,
        contains_any: list[Any] | str | None = None,
        contains_all: list[Any] | str | None = None,
```

The order change might break some user code

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 2]
```diff
@@ -201,34 +206,77 @@ def delete_collections(
...
+                for prop in by_property:
+                    if contains_any:
+                        collection.data.delete_many(where=Filter.by_property(prop).contains_any(contains_any))
+                    elif contains_all:
+                        collection.data.delete_many(where=Filter.by_property(prop).contains_all(contains_all))
+                    else:
+                        collection.data.delete_many(
+                            where=Filter.by_property(prop).like(collection_name + "*")
+                        )
             except Exception as e:
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py]: What would be the expected Exception of these API calls? would it be possible for us to narrow this exception?

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 3]
```diff
@@ -201,34 +206,77 @@ def delete_collections(
...
+    ):
+        """
+        Delete collections based on property names.
+
+        :param collection_name: list of collection names or a single collection name to be deleted.
+        :param by_property: property or list of properties to filter documents by.
+        :param contains_any: list of values or a single value.
+            Deletes collections where any of the specified values are present in the property.
+        :param contains_all: list of values or a single value.
+            Deletes collections  where all specified values are present in the property.
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py]:
```suggestion
            Deletes collections where all specified values are present in the property.
```

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 4]
```diff
@@ -201,34 +206,77 @@ def delete_collections(
...
+        if if_error == "continue":
+            return failed_collection_list
+        return None
+
+    def delete_by_property(
+        self,
+        collection_name: list[str] | str,
+        by_property: list[str] | str,
+        contains_any: list[Any] | str | None = None,
+        contains_all: list[Any] | str | None = None,
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py]: Why do we need to use Any here?

File: airflow/providers/weaviate/hooks/weaviate.py (hunk 1 above)
Comment (harjeevanmaan) [reply airflow/providers/weaviate/hooks/weaviate.py]: Oh sorry! Will commit the change asap

File: airflow/providers/weaviate/hooks/weaviate.py (hunk 2 above)
Comment (harjeevanmaan) [reply airflow/providers/weaviate/hooks/weaviate.py]: @Lee-W Any network exceptions caused by `client.collections.get(collection_name)`
Should I commit?
```
 except requests.exceptions.ConnectionError as e:
        raise e
```

File: airflow/providers/weaviate/hooks/weaviate.py (hunk 4 above)
Comment (harjeevanmaan) [reply airflow/providers/weaviate/hooks/weaviate.py]: It's a silly typo. Apologies

File: airflow/providers/weaviate/hooks/weaviate.py (hunk 2 above)
Comment (Lee-W) [reply airflow/providers/weaviate/hooks/weaviate.py]: Yes, this is perfect!

File: airflow/providers/weaviate/hooks/weaviate.py (hunk 4 above)
Comment (Lee-W) [reply airflow/providers/weaviate/hooks/weaviate.py]: no worries 🙂

File: airflow/providers/weaviate/hooks/weaviate.py [hunk 5]
```diff
@@ -201,35 +207,78 @@ def delete_collections(
...
+                        stop=stop_after_attempt(3),
+                        retry=(
+                            retry_if_exception(lambda exc: check_http_error_is_retryable(exc))
+                            | retry_if_exception_type(REQUESTS_EXCEPTIONS_TYPES)
+                        ),
+                    ):
+                        with attempt:
+                            self.log.info(attempt)
+                            client.collections.delete(collection_name)
+                except Exception as e:
```
Comment (Lee-W) [inline airflow/providers/weaviate/hooks/weaviate.py]: Ah sorry I miss this 🤦‍♂️ I think this needs to be changed as well
```suggestion
                except requests.exceptions.ConnectionError as e:
```

**Review by Lee-W — COMMENTED** [review]
One nitpick left, but almost ready to merge

**Comment by harjeevanmaan** [comment]
Thanks for your guidance

**Review by Lee-W — APPROVED** [review]

**Comment by harjeevanmaan** [comment]
Looking into the test fails

**Comment by Lee-W** [comment]
We'll need to resolve conflicts before the next round of reviewing :)

**Review by dstandish — CHANGES_REQUESTED** [review]
something is screwed up here.  i would start over.

**Comment by harjeevanmaan** [comment]
I agree

**Comment by harjeevanmaan** [comment]
I'm sorry to everyone who got notified. It was a mistake on my end

**Pull request closed without merging**